### ISSUE TITLE
Provide in-product documentation of how uncommitted changes are persisted

### ISSUE THREAD

**Issue by joyceerhl** [microsoft/vscode-remote-repositories-github#263]
Experiment with options, UX sync and additional user studies
        - Use SCM action button API
        - (i) learn more in file explorer title menu for RemoteHub workspaces only, can right click to hide

**Comment by joyceerhl** [comment]
Here are some explorations of how we can clarify that uncommitted changes are stored locally and survive reloads, to reduce user uncertainty:

## Banner in vscode.dev

This is pretty aggressive but ensures users won't miss this info.

i. For PRs when Getting Started isn't displayed
           ![image](https://user-images.githubusercontent.com/30305945/201227465-d467b516-bdbf-4dfb-b809-e6695333730f.png)

   ii. For repos/commits/tags/branches/file deeplinks when Getting Started isn't displayed (e.g. because README or a specific file is already open)
           ![image](https://user-images.githubusercontent.com/30305945/201227485-b0aa4591-f6db-41c0-a364-de5b4c617747.png)

## Getting Started walkthrough

This approach will not help with vscode.dev PRs or file deeplinks because other content is going to be displayed in the editor area. It may help if you come from vscode.dev root instance > Open Remote Repository, GitHub.com, or the Chrome extension.
![image](https://user-images.githubusercontent.com/30305945/201227431-96d19ab2-438c-43be-b500-c11bac598cee.png)
![image](https://user-images.githubusercontent.com/30305945/201227448-94e61f92-5868-4a7e-af4f-a142bf1fde26.png)

## Explorer view in Remote Repositories

For remote repositories only, we could consider showing a view in the explorer which provides context on how uncommitted changes are handled.
<img width="1620" alt="image" src="https://user-images.githubusercontent.com/30305945/201228467-07765cf1-5ae4-4faa-be5e-1ab3c4665699.png">

**Comment by daviddossett** [comment]
@joyceerhl have we received feedback that users are afraid to refresh the browser? Is that a common enough scenario to warrant adding UI?

@digitarald for insights on how successful banners are. My guess is that far fewer people read these than we would think.

**Comment by joyceerhl** [comment]
@daviddossett Yes, this is something that came up repeatedly in the user study that we just ran for Edit Sessions (summary report will be shared out to team early next week). Users expressed confusion about where their changes were stored and were reluctant to refresh the page. They were pleasantly surprised when they learned that uncommitted changes were persisted but didn't expect that to be the case because that's what they've learned from other webpages or other online IDEs. + @stevencl @bamurtaugh for context

**Comment by bamurtaugh** [comment]
+1 to what @joyceerhl mentions, and I just sent the report she mentioned. Data loss is one of users' top concerns, and how changes are persisted throughout vscode.dev and RemoteHub were mysteries to users that we'd really like to clarify for a better first impression and more likely continued use.

> @digitarald for insights on how successful banners are. My guess is that far fewer people read these than we would think.

I do generally think a lot of folks close or miss the top banner, so including info in spots like a getting started walkthrough and RemoteHub view would be helpful. Although they don't have the same top-level benefit as the banner. I'd wonder if we could do anything in the File Explorer, since I'd expect this is another highly-viewed area.

**Comment by stevencl** [comment]
A lot of this is about reducing onboarding anxieties when using vscode.dev for the first time. Not knowing where files are persisted will discourage some people from exploring vscode.dev fully. For example, some people might think that somehow they might end up making changes directly to the repo. Others might think that the repo has been forked/cloned to their own GH account (assuming they are logged in). And others just don't know so they don't even think of making any changes to the code that they are browsing.

If we can reduce this uncertainty and the corresponding anxiety about what will happen to any edits made in vscode.dev, we believe that we will see more people taking more advantage of vscode.dev.

Other brainstorming ideas:
- Pop up dialog on first launch (like the workspace trust dialog) that tells people that any changes they make are stored in a virtual workspace and are not committed to any repository.
- Hover text when hovering over a file in explorer that says "Changes to this and other files are stored in local storage and are not committed to any repository"
- An "Enable editing" button similar to when opening Office docs as attachments that tells people that if they want to edit files in this workspace they should click this - edits are persisted in local storage...
- A notification pop up saying "Creating virtual workspace..." while opening.
- Something like the help panel in Joyce's mockups above but it is positioned at the top of the explorer with a blue background and a close button. Like a popup that remains in place until the user closes it.

**Comment by joyceerhl** [comment]
@bamurtaugh does 'RemoteHub view' refer to the Remote Explorer view, or to a new view that is contributed to the Explorer view or Source Control view?

As another alternative, I think the team has talked about contextual walkthroughs in the past, and IMO the vscode.dev/github case is a great motivating scenario. My hypothesis is that users are more likely to need contextual handholding in vscode.dev than in VS Code for the desktop, because our custom vscode.dev link experiences mean that you can be dropped into a layout containing extensions you didn't explicitly install, files you didn't explicitly open, or views that you didn't explicitly focus. Additionally, if there is default content being displayed, our Getting Started editor with the walkthroughs don't kick in. My hesitation with introducing a `Help` or `About Virtual Workspaces` view in the Explorer view container is that such content only needs to be read once and feels heavyweight to have around for longer. I would prefer a walkthrough experience which allows to display ephemeral context. For example, this is what Figma has:

![image](https://user-images.githubusercontent.com/30305945/202001734-5044ef64-e24c-4c4e-848c-78defcfd6e31.png)

**Comment by bamurtaugh** [comment]
> @bamurtaugh does 'RemoteHub view' refer to the Remote Explorer view, or to a new view that is contributed to the Explorer view or Source Control view?

I meant the Remote Explorer.

> Something like the help panel in Joyce's mockups above but it is positioned at the top of the explorer with a blue background and a close button. Like a popup that remains in place until the user closes it.

@stevencl's idea here ^ is along the lines of what I was thinking, for something more top-level / in the Explorer.

> My hesitation with introducing a Help or About Virtual Workspaces view in the Explorer view container is that such content only needs to be read once and feels heavyweight to have around for longer. I would prefer a walkthrough experience which allows to display ephemeral context.

This makes sense. I think a contextual walkthrough would be great, but I wasn't sure how feasible it was at this point. In the meantime, I was thinking of a note in the Explorer to be similar to the note we pop up in the SCM view's commit box:

![image](https://user-images.githubusercontent.com/25310137/202016392-eabc8833-e8e5-4203-93bc-cd78fee2ad1b.png)

It's a valuable first piece of info, and users can choose not to show it again.

**Comment by digitarald** [comment]
> I do generally think a lot of folks close or miss the top banner, so including info in spots like a getting started walkthrough and RemoteHub view would be helpful. Although they don't have the same top-level benefit as the banner. I'd wonder if we could do anything in the File Explorer, since I'd expect this is another highly-viewed area.

I actually see a lot of users keeping the banner, as it has links that might be useful for later … maybe inducing some FOMO.

How might we make the banner the banner more welcoming and include some of the why-vscode.dev-is-cool.

```
Visual Studio Code (Preview). Anywhere, anytime, entirely in your browser.
Things to try out:
1. Quickly editing local files without installing VS Code. Any edits persist in your browser storage.
2. Share GitHub files, commit changes, and collaborate with Live Share ([How to]).
3. Take any changes and continue on to your desktop VS Code or Codespaces if you need to run, debug, or use the terminal ([How to]).
[Let's go]
```

The best part is that this banner is shown for al users, compared to walkthroughs or any welcome screens.

I wonder how much the information helps in the SCM view, as I understand that user research participants were even unsure about editing in vscode.dev. SCM would only be opened if they have some understanding that editing and committing works.

**Comment by joyceerhl** [comment]
Here's another pass at a view in the explorer container, but further up in the explorer view container and with more connection to Continue On per https://github.com/microsoft/vscode/issues/165406#issuecomment-1315417871. We would only show this on the /github and /azurerepos routes. (On the main vscode.dev site I think we would want to use a Getting Started walkthrough for Remote Repositories--and we could also have one walkthrough per custom route that we ship.)

![image](https://user-images.githubusercontent.com/30305945/202085809-4fa2bca9-c912-4fad-94a9-ee68f5bdfb9f.png)


And a mockup of a toast when the user first adds, deletes, or modifies a file (which I suspect would be missed/ignored):

![image](https://user-images.githubusercontent.com/30305945/202085049-86a9b6fe-cd9d-41d7-8e98-530677b5a1e4.png)

**Comment by stevencl** [comment]
I think the contextual walkthrough is very appealing for a first launch experience. Would we always be able to determine if this is the first time the user has launched vscode.dev?

The notification has higher likelihood of not being noticed although since it is on first launch and the user hasn't started doing anything they may be more likely to pay attention to it than if they had started working and then the notification showed up. I think quite a few people noticed the notification during the study but some said that they had become accustomed to just ignoring them.

Would it be possible to try something like the SCM popup that Brigit showed? The About Remote Repositories panel mockup looks very permanent.

**Comment by joyceerhl** [comment]
Oh I think understand what you're talking about now re: the SCM popover. Something like this? 

![image](https://user-images.githubusercontent.com/30305945/202227424-d6ab3ff0-267b-4e98-ab8e-e6a85f25e3e6.png)

I did some digging and it turns out we already tried the contextual walkthrough idea back when we were first designing the Getting Started walkthroughs: https://github.com/microsoft/vscode/issues/106717#issuecomment-698627820 and https://github.com/microsoft/vscode/issues/106717#issuecomment-790053982 the issue there was that apparently users didn't read most of the content. With a single popover though, maybe we have better odds.

We will need to introduce extension API for this because Remote Repos is an extension and there's no way for extensions to display ephemeral messages in the explorer like this at the moment. The SCM popover has an issue where the user focuses into the input box and then can't see that they have changed files because the popver obscures that content. This popover should only appear when we cannot show context through Getting Started, i.e. the editor will already have content (PR diff, deeplink to file), so we will run into the same issue here.
